beacon-chain/rpc/service.go

```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
 	return &pb.IndexResponse{Index: index}, nil
 }
 
-// LatestAttestation streams the latest processed attestations to the rpc clients.
-func (s *Service) LatestAttestation(req *empty.Empty, stream pb.BeaconService_LatestAttestationServer) error {
-	sub := s.attestationService.IncomingAttestationFeed().Subscribe(s.incomingAttestation)
-	defer sub.Unsubscribe()
+// ValidatorAssignment streams validator assignments every slot to clients that request
+// to watch a subset of public keys in the CrystallizedState's active validator set.
+func (s *Service) ValidatorAssignment(req *pb.ValidatorAssignmentRequest, stream pb.ValidatorService_ValidatorAssignmentServer) error {
+	// If the genesis time was at 12:00:00PM a if the SlotDuration is set to
+	// 8 seconds. We can accomplish this by using utils.WaitUntilTimestamp
+	// and passing in the desired slot duration as an argument.
+	//
+	// Instead of utilizing params.SlotDuration, we utilize a property of
+	// RPC service struct called slotAlignmentDuration. This value
+	// can be set to 0 seconds as a parameter in tests.
+	// Otherwise, tests would sleep.
+	utils.WaitUntilTimestamp(s.slotAlignmentDuration)
+
+	// Right after the timestamps are aligned, we start a new ticker
+	// that fires exactly every params.SlotDuration*time.Second. This ensures
+	// a consistent "heartbeat" to keep track of every beacon slot.
+	tickerChan := time.NewTicker(s.slotAlignmentDuration).C
+	return s.streamValidators(req, stream, utils.CurrentBeaconSlot(), tickerChan, s.ctx.Done())
+}
 
+// streamValidators is a wrapper around the internal logic for streaming validator assignments
+// to clients. It is abstracted into this function for easier tests.
+func (s *Service) streamValidators(
+	req *pb.ValidatorAssignmentRequest,
+	stream pb.ValidatorService_ValidatorAssignmentServer,
+	currentSlot uint64,
+	tickerChan <-chan time.Time,
+	done <-chan struct{}) error {
 	for {
 		select {
-		case attestation := <-s.incomingAttestation:
-			log.Info("Sending attestation to RPC clients")
-			if err := stream.Send(attestation); err != nil {
+		case <-done:
+			log.Debug("ValidatorAssignment stream closed, exiting goroutine")
+			return nil
+		case <-tickerChan:
+
+			if len(req.GetPublicKeys()) == 0 {
+				return errors.New("no public keys specified in request")
+			}
+
+			log.WithField("slotNumber", currentSlot).Debug("Sending validator assignment to RPC clients")
+			// NOTE: this set will not change, so we can just cache the previous one
+
+			// We get the current crystallized state.
+			cState := s.chainService.CurrentCrystallizedState()
+
+			// Next, for each public key in the request, we build
+			// up an array of assignments.
+			assignments := []*pb.ValidatorAssignmentResponse_Assignment{}
+
+			for _, val := range req.GetPublicKeys() {
```

Comment:
This doesn't handle the case where the client requested all assignment updates.

Can you add that logic?